Docs: Add docs for mounting inside the play function

kasperpeulen opened the pull request:
Closes #

<!-- If your PR is related to an issue, provide the number(s) above; if it resolves multiple issues, be sure to break them up (e.g. "closes #1000, closes #1001"). -->

<!--

Thank you for contributing to Storybook! Please submit all PRs to the `next` branch unless they are specific to the current release. Storybook maintainers cherry-pick bug and documentation fixes into the `main` branch as part of the release process, so you shouldn't need to worry about this. For additional guidance: https://storybook.js.org/docs/contribute

-->

## What I did

Add docs for mounting inside the play function

## Checklist for Contributors

### Testing

<!-- Please check (put an "x" inside the "[ ]") the applicable items below to communicate how to test your changes -->

#### The changes in this PR are covered in the following automated tests:
- [ ] stories
- [ ] unit tests
- [ ] integration tests
- [ ] end-to-end tests

#### Manual testing

_This section is mandatory for all contributions. If you believe no manual test is necessary, please state so explicitly. Thanks!_

<!-- Please include the steps to test your changes here. For example:

1. Run a sandbox for template, e.g. `yarn task --task sandbox --start-from auto --template react-vite/default-ts`
2. Open Storybook in your browser
3. Access X story

-->

### Documentation

<!-- Please check (put an "x" inside the "[ ]") the applicable items below to indicate which documentation has been updated. -->

- [ ] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

- [ ] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli/src/sandbox-templates.ts`
- [ ] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

     - `bug`: Internal changes that fixes incorrect behavior.
     - `maintenance`: User-facing maintenance tasks.
     - `dependencies`: Upgrading (sometimes downgrading) dependencies.
     - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
     - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
     - `documentation`: Documentation **only** changes. Will not show up in release changelog.
     - `feature request`: Introducing a new feature.
     - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
     - `other`: Changes that don't fit in the above categories.
   
   </details>

### 🦋 Canary release

<!-- CANARY_RELEASE_SECTION -->

This PR does not have a canary release associated. You can request a canary release of this pull request by mentioning the `@storybookjs/core` team here.

_core team members can create a canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=<PR_NUMBER>`_

<!-- CANARY_RELEASE_SECTION -->

kasperpeulen committed:
Add docs for mounting inside the play function

yannbf reviewed (approved):
Looks good! I added a few comments, please check them out.
Would be nice to have @kylegach take a look at this after he is back!

kasperpeulen committed:
Update docs/_snippets/mount-basic.md

Co-authored-by: Yann Braga <yannbf@gmail.com>

kasperpeulen committed:
Update docs/_snippets/mount-basic.md

Co-authored-by: Yann Braga <yannbf@gmail.com>

kasperpeulen committed:
Update docs/_snippets/mount-basic.md

Co-authored-by: Yann Braga <yannbf@gmail.com>

kasperpeulen committed:
Update docs/writing-tests/interaction-testing.mdx

Co-authored-by: Yann Braga <yannbf@gmail.com>

kasperpeulen committed:
Update docs/writing-tests/interaction-testing.mdx

Co-authored-by: Yann Braga <yannbf@gmail.com>

kasperpeulen committed:
Update docs/writing-tests/interaction-testing.mdx

Co-authored-by: Yann Braga <yannbf@gmail.com>

kasperpeulen commented on docs/_snippets/mount-advanced.md:
```diff
@@ -0,0 +1,95 @@
+    // For example a just generated UUID
+    await mount(<Page {...args} params={{ id: String(note.id) }} />);
+
+    await userEvent.click(await canvas.findByRole('menuitem', { name: /login to add/i }));
```
I will get the canvas from the mount return instead

kasperpeulen committed:
Get the canvas from the mount return

kasperpeulen committed:
Merge remote-tracking branch 'origin/kasper/mount-docs' into kasper/mount-docs

kasperpeulen committed:
Address feedback

kasperpeulen committed:
Merge remote-tracking branch 'refs/remotes/origin/next' into kasper/mount-docs

kasperpeulen committed:
Improve message

kasperpeulen commented on docs/writing-tests/interaction-testing.mdx:
```diff
@@ -56,24 +56,63 @@ Once the story loads in the UI, it simulates the user's behavior and verifies th
+You can do this by adding an asynchronous `beforeEach` function to the meta.
+
+You return a cleanup function from the `beforeEach` function, which will run **after** each story, when the story is remounted or navigated away from.
+You don't need this if you reset the state to a default in the preview file,
```
tried to reword it

kasperpeulen commented on docs/writing-tests/interaction-testing.mdx:
```diff
@@ -56,24 +56,63 @@ Once the story loads in the UI, it simulates the user's behavior and verifies th
 
+### Pass data created in play to the component
+
+When you create data inside the play function, you can pass it to the component in the following way:
```
I added the callout, but I don't understand what you mean with:
> if users are using the mount utility and want to have a custom render function, they can achieve by XYZ

yannbf replied:
What you did is good already!

kasperpeulen committed:
Merge remote-tracking branch 'refs/remotes/origin/next' into kasper/mount-docs

kasperpeulen committed:
Align error message with docs

yannbf commented on code/core/src/preview-errors.ts:
```diff
@@ -221,16 +221,19 @@ export class MountMustBeDestructuredError extends StorybookError {
       message: dedent`
       
       To use mount in the play function, you must use object destructuring, e.g. play: ({ mount }) => {}.
+      This makes sure that storybook won't start rendering before the play function begins.
```
```suggestion
      This makes sure that Storybook won't start rendering your story before the play function begins.
```

yannbf commented on code/core/src/preview-errors.ts:
```diff
@@ -221,16 +221,19 @@ export class MountMustBeDestructuredError extends StorybookError {
+          Ensure that your builder is configured to transpile no earlier than ES2017.
+          
+          This feature is not available:
+          - When you have configured a storybook framework to transpile destructure statements or async/await.
```
```suggestion
          - When you have configured a Storybook framework to transpile destructure statements or async/await.
```

yannbf commented on docs/api/doc-blocks/doc-block-story.mdx:
```diff
@@ -76,6 +76,10 @@ Because all stories render simultaneously in docs entries, play functions can pe
 However, if you know your play function is “safe” to run in docs, you can use this prop to run it automatically.
 
+<Callout variant="info">
+  When the play functions uses `mount`, the story won't be rendered in docs. You must set autoplay to true for such stories if you want to use them in docs.
```
Can you add a hyperlink in `autoplay` which points to the section of that prop? So people know what to do

kasperpeulen replied:
This is the section of `autoplay` :)

yannbf commented on docs/writing-tests/interaction-testing.mdx:
```diff
@@ -56,24 +56,79 @@ Once the story loads in the UI, it simulates the user's behavior and verifies th
+
+<Callout variant="warning">
+  You *must* destructure the mount property from the context.
+  This makes sure that storybook won't start rendering before the play function begins.
```
```suggestion
  This makes sure that Storybook won't start rendering the story before the play function begins.
```

yannbf commented on docs/writing-tests/interaction-testing.mdx:
```diff
@@ -56,24 +56,79 @@ Once the story loads in the UI, it simulates the user's behavior and verifies th
+
+  This feature is not available:
+
+  - When you have configured a storybook framework to transpile destructure statements or async/await.
```
```suggestion
  - When you have configured a Storybook framework to transpile destructure statements or async/await.
```

yannbf reviewed (approved):
Changes look good but I'd love for @kylegach to take a pass at them

kasperpeulen committed:
Update docs/writing-tests/interaction-testing.mdx

Co-authored-by: Yann Braga <yannbf@gmail.com>

kasperpeulen committed:
Update docs/writing-tests/interaction-testing.mdx

Co-authored-by: Yann Braga <yannbf@gmail.com>

kasperpeulen committed:
Update code/core/src/preview-errors.ts

Co-authored-by: Yann Braga <yannbf@gmail.com>

kasperpeulen committed:
Update code/core/src/preview-errors.ts

Co-authored-by: Yann Braga <yannbf@gmail.com>

shilman commented on docs/api/doc-blocks/doc-block-story.mdx:
```diff
@@ -76,6 +76,10 @@ Because all stories render simultaneously in docs entries, play functions can pe
 However, if you know your play function is “safe” to run in docs, you can use this prop to run it automatically.
 
+<Callout variant="info">
+  When the play functions uses `mount`, the story won't be rendered in docs. You must set autoplay to true for such stories if you want to use them in docs.
```
```suggestion
  When a story's play function uses `mount`, the story won't be rendered in docs. You must set autoplay to true for such stories if you want to use them in docs.
```

shilman replied:
Also you should link to the `autoplay` docs here

kylegach replied:
This is under the `autoplay` API reference heading.

shilman commented on docs/writing-tests/interaction-testing.mdx:
```diff
@@ -56,24 +56,79 @@ Once the story loads in the UI, it simulates the user's behavior and verifies th
+You can do this by adding an asynchronous `beforeEach` function to the component meta.
+
+You can return a cleanup function from the `beforeEach` function, which will run **after** each story, when the story is remounted or navigated away from.
+You don't need to add specific cleanup in your component meta `beforeEach`, if you already reset the state to a default in a preview file `beforeEach`.
```
Extra newline before this sentence?

valentinpalkovic commented:
I think we should call out that when the `mount` function is used to actually pass in a new JSXElement with modified arguments, these are NOT synced to the controls panel. Is this a limitation we want to document or actually a bug?

kasperpeulen commented:
@valentinpalkovic A custom `render` function has the exact same limitation right? 

The mount args are really implemented as lazy `render` function, so every limitation with the render function will also apply to the mount args.

valentinpalkovic commented:
That's correct. Still, this limitation should be highlighted and mentioned.

kasperpeulen commented:
@valentinpalkovic But do you mean in both the render and mount docs then?

kylegach committed:
Updates

- Prose tweaks, adding links for context
- Use `If` blocks for conditional content
- Restructure content to be story, component, project-ordered
- Update headings to be problem-oriented

kylegach commented on docs/writing-tests/interaction-testing.mdx:
```diff
@@ -56,24 +56,71 @@ Once the story loads in the UI, it simulates the user's behavior and verifies th
 <Video src="../_assets/writing-tests/addon-interaction-example-optimized.mp4" />
 
-### Run code before each test
+<If notRenderer="angular">
```
I opted to conditionally render this content so that Angular viewers don't even see this feature they cannot use.

But should we render fallback content showing them how to use story-level `beforeEach`? @kasperpeulen

kasperpeulen replied:
I like this! Yes, `beforeEach` should be shown for Angular

kylegach commented on docs/writing-tests/interaction-testing.mdx:
```diff
@@ -56,24 +56,71 @@ Once the story loads in the UI, it simulates the user's behavior and verifies th
+  2. Your Storybook framework or builder must be configured to transpile to ES2017 or newer. This is because destructuring statements and async/await usages are otherwise transpiled away, which prevents Storybook from recognizing your usage of `mount`.
+</Callout>
+
+<If renderer={['react', 'vue', 'svelte']}>
```
Given that the snippets are only available for these renderers, it seemed like the feature itself only worked for these, so I conditionally rendered this section.

Is there appropriate fallback content for the others? @kasperpeulen

kasperpeulen replied:
The fallback is kind of the `basic-mount` example, which will be displayed above this, so I think this is fine!

kylegach commented on docs/writing-tests/interaction-testing.mdx:
```diff
@@ -56,24 +56,71 @@ Once the story loads in the UI, it simulates the user's behavior and verifies th
+
+<CodeSnippets path="mount-basic.md" />
+
+<Callout variant="warning">
+  There are two requirements to use the `mount` function:
+
+  1. You *must* destructure the mount property from the `context` (the argument passed to your play function). This makes sure that Storybook does not start rendering the story before the play function begins.
+  2. Your Storybook framework or builder must be configured to transpile to ES2017 or newer. This is because destructuring statements and async/await usages are otherwise transpiled away, which prevents Storybook from recognizing your usage of `mount`.
+</Callout>
```
If you like the way I reworded this, I suggest updating the error message changed in the PR to match.

kasperpeulen replied:
I like this, but I feel that the error message in Storybook should still explicitly point out Angular.

kasperpeulen committed:
Add note about disabling controls for async args and improve error message

kylegach committed:
Add Angular fallback guidance

kylegach committed:
Merge branch 'next' into kasper/mount-docs

kylegach committed:
Add `beforeAll` section

kylegach committed:
Merge branch 'next' into kasper/mount-docs
